### PR TITLE
Optimize load

### PR DESCRIPTION
**Pull request by jreidinger**
small step for human, but big step for yast

This optimization is done based on performance testing of big hosts file.

result of performance_test script that I also added in this commit:

previous:
```
time ruby performance_test/augeas_loader_test.rb 

real	0m14.590s
user	0m14.584s
sys	0m0.004s
```

after keys cache:
```
time ruby performance_test/augeas_loader_test.rb 

real	0m7.484s
user	0m7.472s
sys	0m0.008s
```

### CODE REVIEW

File: lib/cfa/augeas_parser.rb [hunk 1]
```diff
@@ -263,3 +268,50 @@ def report_error(aug)
     end
   end
 end
+
+# Cache that holds all avaiable keys in augeas tree. It is used to
+# prevent too many aug.match calls which are expensive.
+class AugeasKeysCache
+  STORE_PREFIX = "/store".freeze
+  STORE_LEN = 6
```
Comment (lslezak) [inline lib/cfa/augeas_parser.rb]: `STORE_LEN = STORE_PREFIX.size` is much better.

File: lib/cfa/augeas_parser.rb [hunk 2]
```diff
@@ -263,3 +268,50 @@ def report_error(aug)
...
+  STORE_LEN_1 = 7
```
Comment (lslezak) [inline lib/cfa/augeas_parser.rb]: And `STORE_LEN_1 = STORE_LEN + 1` is way better...

File: lib/cfa/augeas_parser.rb [hunk 3]
```diff
@@ -178,16 +177,20 @@ def report_error(aug)
...
+      key = aug_key[(prefix.size + 1)..-1]
+      key.end_with?("]") ? key.sub(/\[\d+\]$/, "[]") : key
     end
 
-    def load_value(aug, aug_key)
-      nested = !aug.match("#{aug_key}/*").empty?
+    def load_value(aug, aug_key, keys_cache)
+      subkeys = keys_cache.keys_for_prefix(aug_key)
+
+      nested = !subkeys.empty?
```
Comment (lslezak) [inline lib/cfa/augeas_parser.rb]: NP: I'd move the initialization just before where it is used.

File: lib/cfa/augeas_parser.rb [hunk 4]
```diff
@@ -263,3 +268,50 @@ def report_error(aug)
...
+# prevent too many aug.match calls which are expensive.
+class AugeasKeysCache
+  STORE_PREFIX = "/store".freeze
+  STORE_LEN = 6
+  STORE_LEN_1 = 7
+
+  def initialize(aug)
+    fill_cache(aug)
+  end
+
```
Comment (lslezak) [inline lib/cfa/augeas_parser.rb]: I miss some documentation comments in this file...

**Comment by lslezak** [comment]
BTW do we need a ~400KB testing file in Git? There are just ~25 unique lines + 10000 repeated lines. Could we build the testing file before running the test?

Something like:
```ruby
header = "..." # probably a HEREDOC...
data_line = "10.254.128.01   pepa1.labs.suse.cz pepa1"
data = ""
10_000.times { data += data_line }

File.write(testing_file, header + data)
```

**Comment by jreidinger** [comment]
@lslezak what is benefit? I do not want to affect my performance test by any kind of generation, so it is static file, easy to measure. I think 400kb file is not so horrible.

**Comment by lslezak** [comment]
I just found out that the file is not packaged into .gem or .rpm so then it's not critical.

**Pull request merged**